python_modules/libraries/dagster-components/dagster_components/core/component_decl_builder.py

```diff
@@ -11,6 +13,21 @@
 class ComponentFileModel(BaseModel):
     type: str
     params: Optional[Mapping[str, Any]] = None
+    _source_position_tree: Optional[SourcePositionTree] = None
+
+    @property
+    def source_position_tree(self) -> Optional[SourcePositionTree]:
+        return self._source_position_tree
+
+    @staticmethod
+    def from_file(contents: str, filepath: str) -> "ComponentFileModel":
+        parsed = parse_yaml_with_source_positions(contents, filepath)
+        obj = _parse_and_populate_model_with_annotated_errors(
+            cls=ComponentFileModel, obj_parse_root=parsed, obj_key_path_prefix=[]
+        )
+
+        obj._source_position_tree = parsed.source_position_tree  # noqa: SLF001
```

Comment:
We should create a new non-pydantic class that represents the File and all metadata about it.

Reply:
need code smell article for this one :-)

Reply:
Updated to construct a `ComponentFileModel` subclass which includes source data, this lets us have `ComponentFileModel`s with source data and without (e.g. not populated from yaml, as we do in some tests and potentially other codepaths in the future)

Reply:
I don't think this should be a subclass. This is a case where composition > inheritance.

Reply:
I actually think this makes some sense to put as an optional field on `YamlComponentDecl`, which is already a representation of the file and its metadata (e.g. just path, right now). Updated to do that, which doesn't require much surgery.